### PR TITLE
fix(tariscript): multisig ordered signatures and pubkeys

### LINKED ISSUE #5806: Malicious script makes validators waste computations
An attacker can force the CHECK_MULTISIG family of opcodes to be always worst
case by ordering the public keys and their signatures in a way that maximizes the
amount of iterations the nested for-loop must do in order to verify the script.
By putting the public key of the signers and their signatures in reverse order (ie:
the signer for the first signature is the last public key, the second is the second to
last, and so on), an attacker makes all validators waste resources. This is not
reflected in the protocol fees. It is worth keeping in mind that verifying signatures
is an expensive process for validators.

### PR DESCRIPTION
**Pull request by brianp**
Description
---
Update the multisig op code checks to care about the order of keys and signatures. This is to assist with reducing the amount of iterations and processing a node is required to perform to validate a collection.

Motivation and Context
---
Low/Minor audit issue to reduce processing for nodes.

Closes: #5806 

How Has This Been Tested?
---
Test suite / CI

What process can a PR reviewer use to test or verify this change?
---
Look at the test additions. They make it more clear what the effects of the changes are.

Breaking Changes
---

- [x] None
- [ ] Requires data directory on base node to be deleted
- [ ] Requires hard fork
- [ ] Other - Please specify

### CODE REVIEW

File: infrastructure/tari_script/src/script.rs [hunk 1]
```diff
@@ -591,6 +591,7 @@ impl TariScript {
     ///
     /// Notes:
     /// * The _m_ signatures are expected to be the top _m_ items on the stack.
+    /// * The public keys and signatures must both be ordered.
     /// * Every public key can be used AT MOST once.
```
Comment (AaronFeickert) [inline infrastructure/tari_script/src/script.rs]:
```suggestion
    /// * The list may contain duplicate keys, but each occurrence of a public key may be used AT MOST once.
```

File: infrastructure/tari_script/src/script.rs [hunk 2]
```diff
@@ -591,6 +591,7 @@ impl TariScript {
     ///
     /// Notes:
     /// * The _m_ signatures are expected to be the top _m_ items on the stack.
+    /// * The public keys and signatures must both be ordered.
```
Comment (AaronFeickert) [inline infrastructure/tari_script/src/script.rs]:
```suggestion
    /// * The ordering of signatures on the stack MUST match the relative ordering of the corresponding public keys.
```

File: infrastructure/tari_script/src/script.rs (hunk 1 above)
Comment (AaronFeickert) [reply infrastructure/tari_script/src/script.rs]: Should also include a comment that signatures are required to be unique, a case which could legitimately arise on a repeated public key.

File: infrastructure/tari_script/src/script.rs [hunk 3]
```diff
@@ -616,18 +617,24 @@ impl TariScript {
...
-            for (i, pk) in public_keys.iter().enumerate() {
-                if !sig_set.contains(s) && !key_signed[i] && s.verify_raw_canonical(pk, &message) {
-                    // This prevents Alice creating 2 different sigs against her public key
-                    key_signed[i] = true;
+            if pub_keys.len() == 0 {
+                return Ok(None);
+            }
+
+            for pk in pub_keys.by_ref() {
+                if !sig_set.contains(s) && s.verify_raw_canonical(pk, &message) {
```
Comment (AaronFeickert) [inline infrastructure/tari_script/src/script.rs]: I think you can move `!sig_set.contains(s)` outside of (and immediately prior to) the inner `for` loop to save on hash lookups. This means there is a single check for signature duplication. If the signature is verified and it is added to the hash set, the `break` will short-circuit subsequent lookups anyway in favor of the single check after the loop.

Comment (AaronFeickert) [reply infrastructure/tari_script/src/script.rs]: In fact, I think such a move improves clarity of intent as well. It means you assert that a signature initially hasn't been verified already (or else it's a duplicate and the operation fails), and then assert that the signature is eventually verified (or else the operation fails).

**Review by AaronFeickert — CHANGES_REQUESTED** [review]

File: infrastructure/tari_script/src/script.rs (hunk 3 above)
Comment (AaronFeickert) [reply infrastructure/tari_script/src/script.rs]: Tested this change locally, and all tests pass.

Comment (AaronFeickert) [inline infrastructure/tari_script/src/script.rs]: The use of `verify_raw_canonical` means that `message` _must_ safely bind the public key. In this case, it is identical across the full set of allowed public keys, which technically breaks Fiat-Shamir. In practice, it means that a forger would need to target any one of the public keys, the complexity of which depends on the number of public keys in the set.

**Review by AaronFeickert — CHANGES_REQUESTED** [review]

File: infrastructure/tari_script/src/script.rs (hunk 3 above)
Comment (AaronFeickert) [reply infrastructure/tari_script/src/script.rs]: See #5817 for a related issue.

**Review by SWvheerden — APPROVED** [review]

**Pull request merged**